Allow cfg-attributes in where clauses (#3399)
rust-lang/rfcs: text/3399-cfg-attribute-in-where.md

```rust
#[cfg(something)]
trait Dummy: SomeRequirement {}
#[cfg(something)]
impl<T: SomeRequirement> Dummy for T {}
#[cfg(not(something))]
trait Dummy {}
#[cfg(not(something))]
impl<T> Dummy for T {}

impl<T: Dummy> SomeTrait<T> for Thing {}
```

However, this boilerplate does not grow well for multiple conditionally-compiled requirements, becoming rather soupy even at N = 2:

```rust
#[cfg(something_a)]
trait DummyA: SomeRequirementA {}
#[cfg(something_a)]
impl<T: SomeRequirementA> DummyA for T {}
#[cfg(not(something_a))]
trait DummyA {}
#[cfg(not(something_a))]
impl<T> DummyA for T {}

#[cfg(something_b)]
trait DummyB: SomeRequirementB {}
#[cfg(something_b)]
impl<T: SomeRequirementB> DummyB for T {}
#[cfg(not(something_b))]
trait DummyB {}
#[cfg(not(something_b))]
impl<T> DummyB for T {}

impl<T: DummyA + DummyB> SomeTrait<T> for Thing {}
```

Other alternative ways of achieving this also exist, but are typically macro heavy and difficult to implement or check. Importantly, this 
functionality already exists in the language, but quickly grows out of reasonable scope to ergonomically implement.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> Where clauses can use cfg-attributes on individual bounds, like so:

Reviewer: The next example explicitly says "or on functions", but we don't say "on traits" here. 

```suggestion
Where clauses can use cfg-attributes on individual trait bounds, like so:
```

```rust
impl<T> SomeTrait<T> for Thing
where
    #[cfg(something_a)] T: SomeRequirementA,
    #[cfg(something_b)] T: SomeRequirementB,
{}
```
or on functions, including multiple cfg-attributes on a single bound:
```rust
fn some_function<T>(val: &T)
where
    #[cfg(something_a)] 
    T: SomeRequirementA,
    #[cfg(something_b)] 
    #[cfg(not(something_a))] 
    #[cfg(target_os(some_os))] 
    T: SomeRequirementB,
{}

> ```

Reviewer 2: Just something to think about, `rustfmt` is supposed to be able to format the `where` clause somehow, do we expect it to (try to) put the attribute on the same line, or would it always prefer the attribute on separate lines?

Reviewer: I think it would be clearer on separate lines, like existing attributes.

But we're encouraged to defer formatting decisions until after the initial nightly implementation PR:
https://github.com/rust-lang/style-team/blob/main/nightly-style-procedure.md

Reviewer 3: I went ahead and added this as an unresolved question: 57f1df98977b93a0b4fa8ec0ef89b101eee83b7b

> and in other situations where where clauses apply.

Reviewer: It could help readability to clearly distinguish between "where" (English word) and `where` (Rust keyword), because otherwise this could look like a repeated word typo.

```suggestion
and in other situations where `where` clauses apply.
```

During compilation, all cfg-attributes on a where bound are evaluated. If the evaluation result is false, then the bound in question is not
compiled and the bound does not apply to the given type. This may cause errors if code that relies on those bounds is not itself also 
conditionally compiled. For anyone familiar with cfg-attributes already, this should behave similarly to how they are used in, say, struct 
fields or on function signatures.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

In positions that accept where clauses, such as trait implementations and function signatures, individual clauses can now be decorated with 

> cfg-attributes. The cfg-attribute must be on the left hand of the colon (e.g. `#[cfg(...]) T: Foo` rather than `T: #[cfg(...)] Foo`) and 

Reviewer 4: ```suggestion
cfg-attributes. The cfg-attribute must be on the left hand of the colon (e.g. `#[cfg(...)] T: Foo` rather than `T: #[cfg(...)] Foo`) and 
```

> apply for that one bound, up to the comma or end of the where section. Each bound collection will be conditionally compiled depending on the 

Reviewer: "apply for" can mean "ask for" or "is active until". But "applies to" almost always means "is active until".

```suggestion
applies to that one bound, up to the comma or end of the where section. Each bound collection will be conditionally compiled depending on the 
```

builds with different sets of functionality (e.g. client, server, editor, demo, etc.). It would be useful to support cfg-attributes 
directly here without requiring workarounds to achieve this functionality. Macros, proc macros, and so on are also ways to conditionally 
compile where clauses, but these also introduce at least one level of obfuscation from the core goal. Finally, traits can be wholly 
duplicated under different cfg-attributes, but this scales poorly with both the size and intricacy of the trait and the number of 
interacting attributes (which may grow combinatorically), and can introduce a maintenance burden from repeated code.

# Prior art
[prior-art]: #prior-art

I'm not aware of any prior work in adding this to the language. Languages with preprocessors could support this with something like:

```rust
impl<T> SomeTrait<T> for Thing
where
#ifdef SOMETHING_A
    T: SomeRequirementA
#endif
{}
```
but that's not the way I would expect Rust to provide this kind of functionality.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> In theory, I don't see any harm in cfg-attributes decorating individual bounds on the right hand side of the colon. Is it worth adding that

Reviewer: It could help in scenarios where the right-hand side is complex, for example:
`T: Iterator<Item = (U #[cfg(send_bounds)] + Send)>`

But in that case I think a conditionally-compiled type alias is the way to go. (As you can see, the syntax is already a bit weird.)

This is already what many crates do for types like `BoxError`.

Author: That's fair. I don't have a strong opinion on that either way. What I think this would at already enable locally is:

```rust
where
    T: Iterator<Item = U>,
    #[cfg(send_bounds)]
    <T as Iterator>::Item: Send,
```

which would be enough of an ergonomics improvement on its own without needing further complication.

Reviewer: Yes, that seems much clearer.

If users try to add a where bound on the right, the compiler might want to suggest adding it on the left using a fully qualified path instead. But I'm not sure how easy it is to detect those scenarios.
